src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1023,12 +1023,22 @@ public static List<TSource> ToList<TSource>(this IEnumerable<TSource> source)
 
         public static Dictionary<TKey, TSource> ToDictionary<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector)
         {
-            return ToDictionary<TSource, TKey, TSource>(source, keySelector, IdentityFunction<TSource>.Instance, null);
+            return ToDictionary<TSource, TKey>(source, keySelector, null);
         }
 
         public static Dictionary<TKey, TSource> ToDictionary<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey> comparer)
         {
-            return ToDictionary<TSource, TKey, TSource>(source, keySelector, IdentityFunction<TSource>.Instance, comparer);
+            if (source == null) throw Error.ArgumentNull("source");
+            if (keySelector == null) throw Error.ArgumentNull("keySelector");
+
+            int capacity = 0;
+
+            if (source is TSource[]) capacity = (source as TSource[]).Length; // Cast to Array is much faster than cast to ICollection
+            else if (source is ICollection<TSource>) capacity = (source as ICollection<TSource>).Count;
```

Comment:
If the source isn't an array or collection, we're paying for two casts.  Since an array is a collection, would it make sense to only pay for one cast check in the non-collection case? e.g.
```C#
ICollection<TSource> collection = source as ICollection<TSource>;
if (collection != null)
{
    T[] array = source as T[];
    capacity = array != null ? array.Length : collection.Count;
}
...
```
?